src/main/scala/firrtl/options/Phase.scala

```diff
@@ -8,6 +8,32 @@ import logger.LazyLogging
 
 import scala.collection.mutable.LinkedHashSet
 
+object DependencyList {
+  def apply[A](dependencies: DependencyID[A]*): Seq[DependencyID[A]] = {
+    dependencies.toSeq
+  }
+}
+
+case class DependencyID[A](id: Either[Class[_ <: A], A with Singleton]) {
```

Comment:
Does it make more sense to have a `case class` of `Either` or should `DependencyID` be a trait with two concrete subclasses, one for `Class[_ <: A]` and one for `A with Singleton`?

Reply:
That's a good point. I used `Either` to make the "one or the other" aspect a bit more self-documenting, but case classes do have some nice benefits:

- Dynamic dispatch instead of `Left`/`Right`-casing for `getObject`
- Consistency with pattern that people actually see in the rest of FIRRTL

I guess it also comes down to whether anyone actually seeing `Either` could ever be usefully self-documenting or whether it would just confuse people.